service/s3.cc

```diff
@@ -2705,45 +2709,31 @@ void registerS3Buckets(const std::string & accessKeyId,
                        const std::string & protocol,
                        const std::string & serviceUri)
 {
-    std::unique_lock<std::recursive_mutex> guard(s3BucketsLock);
-    int bucketCount(0);
-    auto api = std::make_shared<S3Api>(accessKeyId, accessKey,
-                                       bandwidthToServiceMbps,
-                                       protocol, serviceUri);
-
-    auto onBucket = [&] (const std::string & bucketName)
-        {
-            //cerr << "got bucket " << bucketName << endl;
-
-            S3BucketInfo info;
-            info.s3Bucket = bucketName;
-            info.api = api;
-            s3Buckets[bucketName] = info;
-            bucketCount++;
-
-            return true;
-        };
-
-    api->forEachBucket(onBucket);
-
-    if (bucketCount == 0) {
-        cerr << "registerS3Buckets: no bucket registered\n";
-    }
+    std::unique_lock<std::mutex> guard(keyInfoLock);
+    globalKeyInfo = KeyInfo(accessKeyId, accessKey,
+                            bandwidthToServiceMbps,
+                            protocol, serviceUri);
 }
 
 std::shared_ptr<S3Api> getS3ApiForBucket(const std::string & bucketName)
 {
-    std::unique_lock<std::recursive_mutex> guard(s3BucketsLock);
-    auto it = s3Buckets.find(bucketName);
-    if (it == s3Buckets.end()) {
-        // On demand, load up the configuration file before we fail
-        registerDefaultBuckets();
-        it = s3Buckets.find(bucketName);
+    std::unique_lock<std::mutex> guard(keyInfoLock);
+
+    KeyInfo keyInfo = globalKeyInfo;
+    auto it = specificKeyInfos.find(bucketName);
+    if (it != specificKeyInfos.end()) {
+        keyInfo = it->second;
     }
-    if (it == s3Buckets.end()) {
-        throw ML::Exception("unregistered s3 bucket " + bucketName);
+
+    if (keyInfo.accessKeyId.empty()) {
+        throw ML::Exception("no s3 credentials registered for bucket " + bucketName);
     }
-    return it->second.api;
+
+    return make_shared<S3Api>(keyInfo.accessKeyId,
```

Comment:
With this implementation (HttpClient based), the connection pool is kept elsewhere. Now, I realise that I might indeed use S3Api directly, since it's basically holding the same data.